fsw/cfe-core/src/sb/cfe_sb_api.c

```diff
@@ -1397,6 +1397,7 @@ int32  CFE_SB_TransmitMsg(CFE_MSG_Message_t *MsgPtr, bool IncrementSequenceCount
 
     PendingEventID = 0;
     BufDscPtr = NULL;
+    RouteId = CFE_SBR_INVALID_ROUTE_ID;
```

Comment:
What about having CFE_SB_TransmitMsgValidate actually set RouteId (actually done by the underlying GetRouteId call)?  Then it fixes everywhere the GetRouteId call is used.

Reply:
We can do that too, but I see it as a separate problem ...

1. A caller should never rely on/use output values from a function that returned non-success
2. A function should always set its output buffers to something, even when it returns non-success

I see (1) as the main issue, and (2) as the backup/failsafe.  I have no objection to doing both, I just see it as a secondary concern.  I can add it though.

Reply:
In virtually every case I've seen that the static analyzer or compiler warning for possible uninitialized variable comes up 1 is met by the caller by either returning or checking status prior to using what is returned.  2 then is both good practices and squashes the warning.  Is initializing the variable in this case really required where 1 is already satisfied by logic?

This seems even more useless when the value being initialized is not checked explicitly before used, and would cascade into other issues down the line if there WAS a possible way to get to that code.  Basically, if you are checking for status ==  SUCCESS but depend on RouteId be valid why not check instead (or in addition if both cases need to be true) for RouteId being valid (that also squashes the warning).

Reply:
Note for the case where the code IS checking for RouteId to be valid, I definitely agree explicit initialization is the way to go.  I just struggle with justifying initialization if it's never accessed based on failure logic.  Feels cleaner to me (and would be easier for me to maintain) when initialization is done locally when it matters (and squash errors via setting outputs in the underlying APIs when it doesn't matter).

Reply:
@skliper I added this in commit bf8f481 however I have my concerns/reservations - there is a chance that the original passed-in buffers themselves are NULL, where the status code becomes `CFE_SB_BAD_ARGUMENT` ... this will then try to write a value to the buffer.  Could add NULL checks on each, but then this is just getting overly complicated at that point.

Reply:
Good point, I was thinking it would make sense to set them "inline" where it made sense (and really the getter calls could handle most of it, other then the elements not set yet).  Although I did just hit approve.

Reply:
Alternatively just return for the NULL check failure, and initialize them to known values right after that?  Then you aren't overwritting anything that may have succeeded...

Reply:
Or as an else case on the null check

Reply:
> ...require substantial rework...

It already Is it really that big of an ask?  It checks for one pointer for NULL, seems like either all (and setting defaults only if good) or none would make more sense.  Either they can never be NULL or should be checked, no?  I'm fine with either approach as an internal function (and really don't have a real problem with your current PR), but just checking one pointer seems like a strange pattern to employ.  In the end I don't really care, but I really don't understand the pushback.  Could you provide background for your opposition?

Reply:
Or if:
https://github.com/nasa/cFE/blob/bf8f481e2d87c13e59e2a32827f24dbd13c27394/fsw/cfe-core/src/sb/cfe_sb_api.c#L1536-L1541

Is really the sweet spot for checking, why not just add the else and set the outputs to defaults here instead of at the end?  Then it avoids setting them for the `CFE_SB_BAD_ARGUMENT` case.

Reply:
I mean, anything is possible, if you want to do that.  Each output _could_ be individually checked for NULL and assigned a default if not null.  If it helps meet a checkbox on some review procedure, that's well and good - but it doesn't make sense in the overall design of the software, it is just unnecessary extra complexity for no actual benefit (and cpu cycles burnt in code that actually does execute rather frequently in CFE)

I'm just taking a more holistic view of the software.  This is an internal helper function, called only from the two public "TransmitMsg" variants, in a specific scenario, to consolidate the common logic.

The only reason that `MsgPtr` is explicitly checked is because of the fact that (for some reason?) public APIs are all supposed to accept and detect a NULL pointer input.  This assumes the role of checking that input.

For the other 3 - those are outputs to stack variables.  They are stack variables in the parent - never NULL.  This is not a public API - they aren't going to be randomly called with arbitrary values in a general-purpose sense.  Again, the role of this function is to consolidate common code between CFE_SB_TransmitMsg and CFE_SB_TransmitBuffer, initializing local stack variables on behalf of the caller, such that the code is not repeated between the two.

By definition, any NULL pointer check here is impossible to get at runtime - thereby making it dead code.  The way I see it this is very similar to the recent case in OSAL where a check for an impossible condition was removed due to the code before it.  This is the exact same thing, just in a helper function.  The helper is not invoked with NULL args, so any check for NULL args is immediately dead/unreachable code.

Reply:
It was you're comment about "there is a chance that the original passed-in buffers themselves are NULL, where the status code becomes CFE_SB_BAD_ARGUMENT" that triggered my comments... so can they be NULL or not?  If not, the current implementation is fine (I already approved that one when I first saw it.  If they can be NULL, then check for it.  Alternatively, just set the passed in pointers as the else of the MsgPtr check.  Are all 3 of these possible approaches really not acceptable and the only option is to not initialize these values? I still don't follow why there is any opposition?  Maybe contact me for a chat?

Reply:
Sorry about that - that was only my initial knee-jerk before looking closer and seeing that the function really only does need to handle MsgPtr being null and not the others.  To allow the others to be null would be pointless.

Reply:
Copy.  And I'm fine w/ how it's currently just checking MsgPtr (probably would benefit from a comment to that affect, but implementation is fine) based on the background you provided.